text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,639 @@
+interested in AWS-vended constructs anyway and they will all be implemented
+in TypeScript.
+
+> Should we ever want to open this mechanism up to non-TypeScript libraries, an obvious
+> mechanism is to introduce a well-known metadata tag that we can read a construct's
+> construct identifier from (falling back to the default search algorithm if none
+> is found). It becomes a library author's concern of how to attach the metadata
+> then.
+
+## Metadata resource encoding
+
+Once we have the list of construct identifiers, we need to encode it into the metadata
+resource.
+
+The Metadata resource looks like this:
+
+```yaml
+Resources:
+  CDKMetadata:
+    Type: AWS::CDK::Metadata
+    Properties:
+      Modules: <v1 module information, deprecated by this RFC>
+      Analytics: <v2 construct information, introduced by this RFC>
+```
+
+### Construct list size
+
+The list of construct identifiers we have discovered may be sizeable. A rough
+estimate of the maximum number of constructs used in a stack will be about
+400 (maximum of 200 L1s with a corresponding L2 wrapper). We will therefore
+need to encode roughly 400 strings of the form
+`@aws-solutions-constructs/aws-lambda-sqs.LambdaToSqs@1.63.0`.
+
+The longest overall type in the current monocdk is
+`monocdk-experiment.aws_cognito.CfnUserPoolRiskConfigurationAttachment.CompromisedCredentialsRiskConfigurationTypeProperty`
+(121 characters), the longest construct class name is
+`monocdk-experiment.aws_ecs_patterns.ApplicationMultipleTargetGroupsFargateService`
+(81 characters).
+
+If an identifier is estimated to be a maximum of 100 bytes (which seems
+reasonable be plenty of room) the resulting data set will be at most ~40kB.
+
+We can encode the elements as a list straight in the template, but it will
+take up a lot of space in the template: vertical space while looking at it,
+as well as bytes that count toward the template size (maximum of 460kB).
+
+### Compression
+
+As an experiment, I've drawn 400 random class names from the complete list of class
+names currently in monocdk; that set of class names comes to about 23kB (uncompressed).
+
+Trying various methods of compression on it yields the following (keeping in mind
+that the result of binary compression will need to be base64-encoded to transmit via
+a CloudFormation template):
+
+| Description                | Raw size    | Base64ed size |
+|----------------------------|-------------|---------------|
+| Plaintext                  | 23k         | -             |
+| gzipped                    | 3.6k        | 4.8k          |
+| Zopfli                     | 3.4k        | 4.6k          |
+| bzip2                      | 3.4k        | 4.6k          |
+| Prefix-grouped (plaintext) | 7.4k        | -             |
+| Prefix-grouped gzipped     | 3.0k        | 4.0k          |
+| Bloom Filter (1% error)    | 720b        | 960b          |
+
+Prefix-grouped, gzipped data seems to give the best results that are still
```

Comment:
Assuming the same 400 constructs, this comes to 8% of the template if it were going inline (max 50K) as part of `createStack`, and <1% if the template was sent via S3 (max 450K).

I assume a template with 400 constructs cannot fit into the 50K limit anyway. I did a quick experiment and ~85 construct *instances* were needed to fill up to 50K. The actual construct types will be much lesser. 

The compression reached by prefix-grouped gzipped feels fine to me.